compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -1615,6 +1615,19 @@ impl EnumMemberDescriptionFactory<'ll, 'tcx> {
                 // For MSVC, we will generate a union of two structs, one for the dataful variant and one that just points to
                 // the discriminant field. We also create an enum that contains tag values for the non-dataful variants and
                 // make the discriminant field that type. We then use natvis to render the enum type correctly in Windbg/VS.
+                // This will generate debuginfo roughly equivalent to the following C:
+                // ```c
+                // union <enum name> {
+                //   struct dataful_variant {
+                //     <fields in dataful variant>
+                //   },
+                //   struct discriminant$ {
+                //     enum tag$ {
+                //       <non-dataful variants>
+                //     } discriminant;
+                //   }
+                // }
+                // ```
```

Comment:
Thanks!

So, for decoding, one is supposed to get numeric value of `discriminant$.discriminant` and if it is in the range between the second and the third template parameters (inclusive or exclusive?), display `dataful_variant`, using variant name from the last template parameter; otherwise display variant name corresponding to the discriminant value?

If not too much trouble, could you please also add layout spec for non-niche enums?

Reply:
Yes, that's exactly right!

Done 🙂